design-documents/graph-ql/coverage/catalog/configurable-options-selection.md

```diff
@@ -0,0 +1,205 @@
+## Use cases
+
+### Render configurable option values available for selection on the product page
+
+User navigates to the configurable product page. Option values available for selection are rendered on the page.
+
+```graphql
+{
```

Comment:
I've been looking at this query for a bit, and trying to think through how folks like @magento/pwa-studio-team are going to consume this and render it in a UI.

It _looks_ like the UI will do something like this:

- Loop over `data.products.items[0].configurable_options`
  - For each option, grab `attribute_code`
  - Loop over `configurable_options_selection_metadata.configurable_options_available_for_selection` for the matching `attribute_code` from the out loop

That nested looping is necessary because `configurable_options_selection_metadata` is disconnected from `configurable_options`, and is sort-of weakly referenced via an ID, leaving it to the UI to go find that object.

What would the impact be if we moved `configurable_options_selection_metadata` into `ConfigurableProductOptions`?

```graphql
{
    products(filter: {sku: {eq: "sku-here"}}) {
        ...on ConfigurableProduct {
            configurable_options(selectedOptionValueIDs: [ID!]) {
                attribute_code
                label
                values {
                   id # ID can be checked against `available_value_ids`
                }
                available_value_ids
            }
        }
    }
}
```

Thoughts?

Reply:
As mentioned in other comments above, in this case we are loosing the "Metadata" aspect and extensibility.

My understanding was that on the first request (with no selections made), the client will only be using `configurable_options` field. We can add availability flag to the values there, it makes sense in my opinion.
The subsequent queries only ask for `options_available_for_selection` and probably other metadata like `variant`, `price_range` etc. They already have a list of all options with values from the initial request and do not need to query `configurable_options` anymore.

Reply:
> My understanding was that on the first request (with no selections made), the client will only be using configurable_options field

This is my understanding as well, so far on the same page 👍 

> We can add availability flag to the values there, it makes sense in my opinion

Cool 👍 I like the idea.

> They already have a list of all options with values from the initial request and do not need to query configurable_options anymore.

If some fields aren't necessary on the second request (`attribute_code`/`values`/`label`) the client just won't select those fields. I don't think that's necessarily a reason to split these data types out into separate fields things (`configurable_options` vs `configurable_options_selection_metadata `).

> in this case we are loosing the "Metadata" aspect and extensibility.

It's Monday and I'm not running on all cylinders yet 😅 - can you give me an example of the extensibility we lose this way?

Reply:
Added `is_available_for_selection` flag to option value type.

Regarding extensibility, I mean that the fields like `price_range`, `variant`, `low_stock_status` will not fit into existing `configurable_options` field. While they look really good in `configurable_options_selection_metadata` in my opinion.

Reply:
> While they look really good in configurable_options_selection_metadata in my opinion.

My biggest concern is still around how ergonomic this is for UI developers to consume, which is the APIs primary audience.

In a typical modern front-end, things are broken up by components, and you just describe how you want the UI to look by composing those components.

Let's step a bit back from this specific schema, and talk more in terms of how the UI code would be structured.

In this case, the UI code would ideally just want to loop over the options groups (color/size/etc), and then render the options with that dataset.

```jsx
// Something like this using JSX, but pretty similar in a template-based world like Vue
const options = option_groups.map(option => 
   <OptionGroup
     attrCode={option.attribute_code}
     label={option.label}
     values={option.values}
   />
);
```

Then, the `<OptionGroup />` component would just iterate over `values`, and spit out a list of HTML elements representing all the values in the option group. Ideally, `option.values` here would have data about whether or not the option is selectable, so it can be greyed-out, or sorted to move available options to the front of the list.

With the currently proposed schema, if the UI wanted to do this, it would have to either have to normalize data into a lookup table every time a response comes in, or do something like this:

```jsx
const options = options_groups.map(option =>
   <OptionGroup
     attrCode={option.attribute_code}
     label={option.label}
     values={option.values}
     // This nested loop could be several hundred iterations for each option group,
     // and just isn't super fun to write
     availableValues={options.find(o => o.attribute_code === option.attributeCode)}
   />
);
```

Reply:
Having said all of that, if you can't see an alternative schema design working, we can always start with this and change course or add later if we find usability problems.